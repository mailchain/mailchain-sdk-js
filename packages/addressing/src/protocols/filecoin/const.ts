import { SECP256K1PrivateKey } from '@mailchain/crypto';
import { decodeHex } from '@mailchain/encoding';

// Alice
export const AliceFilStr = 'f410faiikgixcdd7alrsbpunjlen4etkcarbfbghn7wi';
export const AliceFilEthAddressStr = '0x0210a322e218fe05c6417d1a9591bc24d4204425';
export const AliceFilEthPrivateKeyStr = '6c115551accbd74b46401a531af1dd0890034c18137dddab52a53c626e75b051';
export const AliceFilEthPrivateKey = new SECP256K1PrivateKey(decodeHex(AliceFilEthPrivateKeyStr));

// Bob
export const BobFilAddressStr = 'f410f5blfdi7tmto6o2aa2up4vwtdm7qcnaiqpeazuwy';
export const BobFilEthAddressStr = '0xe85651a3f364dde76800d51fcada6367e0268110';
export const BobFilEthPrivateKeyStr = '2c6b29d5545e5804b7fb7e65404782e57dddea84cbcec9b2586add6403808965';
export const BobFilEthPrivateKey = new SECP256K1PrivateKey(decodeHex(BobFilEthPrivateKeyStr));

export enum FilPrefix {
	F4ETHEREUM = 'f410',
	T4ETHEREUM = 't410',
}
