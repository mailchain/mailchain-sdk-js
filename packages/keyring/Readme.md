# Mailchain Keyring

Mailchain is a messaging protocol that lets users communicate across protocols. Using Mailchain you can send messages to any blockchain address on different protocols.

For full usage examples view the [developer docs](https://docs.mailchain.com).

## Installing

Using npm:

```bash
$ npm install @mailchain/keyring
```

Using yarn:

```bash
$ yarn add @mailchain/keyring
```

## Purpose

The keyring package is used by `@mailchain/sdk` to support securely access messaging and account keys.
