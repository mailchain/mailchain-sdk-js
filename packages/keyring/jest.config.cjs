const path = require('path');
const { createConfig } = require('../../../../jest.config.cjs');

module.exports = {
	...createConfig(path.join(__dirname, '..', '..', '..', '..')),
};
