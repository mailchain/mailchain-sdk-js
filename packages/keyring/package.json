{
	"name": "@mailchain/keyring",
	"version": "0.20.0",
	"description": "Mailchain keyring",
	"license": "Apache-2.0",
	"author": {
		"name": "Mailchain",
		"email": "support@mailchain.co",
		"url": "https://mailchain.com/"
	},
	"repository": {
		"type": "git",
		"url": "https://github.com/mailchain/mailchain-sdk-js.git",
		"directory": "keyring"
	},
	"bugs": {
		"url": "https://github.com/mailchain/mailchain-sdk-js/issues"
	},
	"dependencies": {
		"@mailchain/addressing": "0.20.0",
		"@mailchain/crypto": "0.20.0",
		"@mailchain/encoding": "0.20.0"
	},
	"scripts": {
		"build": "cross-env NODE_ENV=production rollup -c --bundleConfigAsCjs",
		"test": "run -T jest",
		"lint:test": "yarn run -T eslint --ext js,ts,tsx src",
		"lint:fix": "yarn run -T eslint --fix --ext js,ts,tsx src"
	}
}
