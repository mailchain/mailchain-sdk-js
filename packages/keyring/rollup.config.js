import { getBundlingConfigs } from '../library.rollup.config';
import pkg from './package.json';

export default getBundlingConfigs('keyring', pkg);
