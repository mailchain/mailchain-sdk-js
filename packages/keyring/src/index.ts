export * from './keyring';
export * from './functions';
