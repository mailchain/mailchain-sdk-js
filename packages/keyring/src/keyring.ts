import {
	ED25519PrivateKey,
	DeriveHardenedKey,
	ED25519ExtendedPrivateKey,
	AsED25519PublicKey,
	AsED25519PrivateKey,
} from '@mailchain/crypto/ed25519';
import { EncodeBase58 } from '@mailchain/encoding/base58';
import { PublicKey, PrivateKey, Encrypter, Decrypter, PublicKeyEncrypter, PublicKeyDecrypter } from '@mailchain/crypto';
import { EncodeHex } from '@mailchain/encoding';
import { protocols } from '@mailchain/internal';
import {
	DERIVATION_PATH_ENCRYPTION_KEY_ROOT,
	DERIVATION_PATH_IDENTITY_KEY_ROOT,
	DERIVATION_PATH_INBOX_ROOT,
	DERIVATION_PATH_USER_PROFILE,
	DERIVATION_PATH_MESSAGING_KEY_ROOT,
} from './constants';

export class KeyRing {
	private readonly _accountIdentityKey: ED25519ExtendedPrivateKey;
	private readonly _rootEncryptionKey: ED25519ExtendedPrivateKey;
	private readonly _userProfileEncryptionKey: ED25519ExtendedPrivateKey;
	private readonly _rootInboxKey: ED25519ExtendedPrivateKey;
	// used to derive messaging keys for all protocol addresses
	private readonly _protocolAddressRootMessagingKey: ED25519ExtendedPrivateKey;
	private readonly _accountMessagingKey: ED25519ExtendedPrivateKey;
	/**
	 *
	 * @param accountKey This key is never stored in the key chain only used to derive other keys.
	 */
	constructor(accountKey: ED25519PrivateKey) {
		this._accountIdentityKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(accountKey),
			DERIVATION_PATH_IDENTITY_KEY_ROOT,
		);

		// used to derive all messaging keys
		const rootMessagingKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(accountKey),
			DERIVATION_PATH_MESSAGING_KEY_ROOT,
		);

		this._protocolAddressRootMessagingKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(rootMessagingKey.PrivateKey),
			1,
		);

		// e.g. bob@mailchain
		const rootAccountMessagingKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(rootMessagingKey.PrivateKey),
			'protocol=mailchain',
		);

		// default to nonce of 1 for now
		this._accountMessagingKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(rootAccountMessagingKey.PrivateKey),
			1,
		);

		this._rootEncryptionKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(accountKey),
			DERIVATION_PATH_ENCRYPTION_KEY_ROOT,
		);
		this._userProfileEncryptionKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(this._rootEncryptionKey.PrivateKey),
			DERIVATION_PATH_USER_PROFILE,
		);
		this._rootInboxKey = DeriveHardenedKey(
			ED25519ExtendedPrivateKey.FromPrivateKey(this._rootEncryptionKey.PrivateKey),
			DERIVATION_PATH_INBOX_ROOT,
		);
	}
	static async Generate(): Promise<KeyRing> {
		// TODO: temporary for testing
		return new this(await ED25519PrivateKey.Generate());
	}

	static FromMnemonic(mnemonic: string, password?: string): KeyRing {
		return new this(ED25519PrivateKey.FromMnemonicPhrase(mnemonic, password));
	}

	static FromPrivateKey(key: ED25519PrivateKey): KeyRing {
		return new this(key);
	}

	createIdentityKeyForPublicKey(key: PublicKey): ED25519PrivateKey {
		return AsED25519PrivateKey(DeriveHardenedKey(this._accountIdentityKey, key.Bytes).PrivateKey);
	}

	createMessagingKeyForAddress(
		address: Uint8Array,
		protocol: protocols.ProtocolType,
		nonce: number,
	): ED25519PrivateKey {
		// all addresses are encoded with hex regardless of protocol to ensure consistency
		const addressKeyRoot = DeriveHardenedKey(
			this._protocolAddressRootMessagingKey,
			`protocol=${protocol},address=${EncodeHex(address)}`,
		);

		// specific for the nonce
		return DeriveHardenedKey(addressKeyRoot, nonce).PrivateKey;
	}

	rootIdentityPublicKey(): PublicKey {
		return AsED25519PublicKey(this._accountIdentityKey.PrivateKey.PublicKey);
	}

	rootEncryptionPublicKey(): PublicKey {
		return AsED25519PublicKey(this._rootEncryptionKey.PrivateKey.PublicKey);
	}

	accountMessagingPublicKey(): PublicKey {
		return AsED25519PublicKey(this._accountMessagingKey.PrivateKey.PublicKey);
	}

	rootInboxKey(): PrivateKey {
		return this._rootInboxKey.PrivateKey;
	}

	EncodedAddress(): string {
		if (this._accountIdentityKey) {
			return EncodeBase58(this._accountIdentityKey.PrivateKey.PublicKey.Bytes);
		}
		return '';
	}

	userProfileCrypto(): {
		encrypter: Encrypter['Encrypt'];
		decrypter: Decrypter['Decrypt'];
	} {
		return {
			encrypter: (input) =>
				PublicKeyEncrypter.FromPublicKey(this._userProfileEncryptionKey.PrivateKey.PublicKey).Encrypt(input),
			decrypter: (input) =>
				PublicKeyDecrypter.FromPrivateKey(this._userProfileEncryptionKey.PrivateKey).Decrypt(input),
		};
	}

	async SignWithIdentityKey(payload: string): Promise<Uint8Array> {
		return this._accountIdentityKey.PrivateKey.Sign(Buffer.from(payload, 'utf8'));
	}

	async SignWithMessageKey(payload: string): Promise<Uint8Array> {
		return this._accountMessagingKey.PrivateKey.Sign(Buffer.from(payload, 'utf8'));
	}
}
