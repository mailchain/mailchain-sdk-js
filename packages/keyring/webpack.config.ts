import { getBrowserLibraryConfig, getNodeLibraryConfig } from '../../scripts/library.webpack.config';
import { dependencies } from './package.json';

const libraryDetails = {
	name: 'keyring',
	type: 'umd',
};

const externals = Object.keys(dependencies)
	.filter((x) => x.startsWith('@mailchain'))
	.reduce((acc, rec) => {
		return { ...acc, [rec]: rec };
	}, {});

const configs = [getNodeLibraryConfig(libraryDetails, externals), getBrowserLibraryConfig(libraryDetails, externals)];
// eslint-disable-next-line import/no-default-export
export default configs;
