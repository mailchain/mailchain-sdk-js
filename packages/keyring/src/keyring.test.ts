import { AliceED25519PrivateKey, BobED25519PrivateKey } from '@mailchain/crypto/ed25519/test.const';
import { ETHEREUM, MAILCHAIN } from '@mailchain/internal/protocols';
import { KeyRing } from './keyring';

describe('keyring', () => {
	it('should assert against previous snapshot outputs from alice keyring', async () => {
		const kr = KeyRing.fromPrivateKey(AliceED25519PrivateKey);

		expect(kr.inboxMessageDateOffset()).toEqual(204954984);
		expect(kr.addressMessagingKey(new Uint8Array([1, 3, 3, 7]), ETHEREUM, 1)).toMatchSnapshot(
			'addressMessagingKey',
		);
		expect(kr.addressMessagingKey(new Uint8Array([1, 3, 3, 7]), MAILCHAIN, 1).publicKey).toEqual(
			kr.accountMessagingKey().publicKey,
		);
		expect(kr.rootInboxKey()).toMatchSnapshot('rootInboxKey');
		expect(kr.rootEncryptionPublicKey()).toMatchSnapshot('rootEncryptionPublicKey');
		expect(kr.accountMessagingKey()).toMatchSnapshot('accountMessagingKey');
		expect(kr.accountMessagingKey().sign(new Uint8Array([1, 3, 3, 7]))).toMatchSnapshot('accountMessagingKey');
		expect(kr.accountIdentityKey()).toMatchSnapshot('accountIdentityKey');
		expect(await kr.accountIdentityKey().sign(new Uint8Array([1, 3, 3, 7]))).toMatchSnapshot('accountIdentityKey');
	});

	it('should assert against previous snapshot outputs from bob keyring', async () => {
		const kr = KeyRing.fromPrivateKey(BobED25519PrivateKey);

		expect(kr.inboxMessageDateOffset()).toEqual(635667296);
		expect(kr.addressMessagingKey(new Uint8Array([1, 3, 3, 7]), ETHEREUM, 1)).toMatchSnapshot(
			'addressMessagingKey',
		);
		expect(kr.addressMessagingKey(new Uint8Array([1, 3, 3, 7]), MAILCHAIN, 1).publicKey).toEqual(
			kr.accountMessagingKey().publicKey,
		);
		expect(kr.rootInboxKey()).toMatchSnapshot('rootInboxKey');
		expect(kr.rootEncryptionPublicKey()).toMatchSnapshot('rootEncryptionPublicKey');
		expect(kr.accountMessagingKey()).toMatchSnapshot('accountMessagingKey');
		expect(kr.accountMessagingKey().sign(new Uint8Array([1, 3, 3, 7]))).toMatchSnapshot('accountMessagingKey');
		expect(kr.accountIdentityKey()).toMatchSnapshot('accountIdentityKey');
		expect(await kr.accountIdentityKey().sign(new Uint8Array([1, 3, 3, 7]))).toMatchSnapshot('accountIdentityKey');
	});

	it('should encrypt and decrypt for user profile data', async () => {
		const kr = KeyRing.fromPrivateKey(AliceED25519PrivateKey);
		const userProfileCrypto = kr.userProfileCrypto();
		const input = new Uint8Array([1, 3, 3, 7]);

		const encrypted = await userProfileCrypto.encrypt(input);
		const decrypted = await userProfileCrypto.decrypt(encrypted);

		expect(decrypted).toEqual(input);
	});
});
