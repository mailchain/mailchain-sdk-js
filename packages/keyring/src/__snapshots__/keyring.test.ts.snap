// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyring should assert against previous snapshot outputs from alice keyring: accountIdentityKey 1`] = `
{
  "curve": "ed25519",
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      55,
      165,
      208,
      63,
      15,
      176,
      134,
      129,
      169,
      191,
      181,
      227,
      13,
      206,
      231,
      167,
      68,
      229,
      165,
      122,
      204,
      67,
      215,
      190,
      151,
      155,
      0,
      166,
      198,
      178,
      77,
      27,
    ],
    "curve": "ed25519",
  },
  "sign": [Function],
}
`;

exports[`keyring should assert against previous snapshot outputs from alice keyring: accountIdentityKey 2`] = `
Uint8Array [
  10,
  21,
  96,
  217,
  214,
  156,
  0,
  226,
  103,
  31,
  188,
  248,
  55,
  167,
  117,
  18,
  118,
  159,
  242,
  82,
  5,
  119,
  46,
  5,
  200,
  163,
  71,
  206,
  46,
  177,
  101,
  132,
  5,
  94,
  145,
  205,
  248,
  202,
  67,
  38,
  207,
  224,
  116,
  79,
  254,
  12,
  226,
  26,
  138,
  59,
  248,
  145,
  40,
  17,
  210,
  67,
  235,
  85,
  46,
  58,
  168,
  225,
  137,
  0,
]
`;

exports[`keyring should assert against previous snapshot outputs from alice keyring: accountMessagingKey 1`] = `
{
  "curve": "ed25519",
  "ecdhDecrypt": [Function],
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      215,
      190,
      70,
      167,
      83,
      44,
      32,
      26,
      113,
      228,
      19,
      133,
      8,
      32,
      97,
      216,
      16,
      67,
      139,
      65,
      185,
      41,
      147,
      108,
      27,
      108,
      75,
      48,
      31,
      33,
      50,
      155,
    ],
    "curve": "ed25519",
  },
  "sign": [Function],
}
`;

exports[`keyring should assert against previous snapshot outputs from alice keyring: accountMessagingKey 2`] = `Promise {}`;

exports[`keyring should assert against previous snapshot outputs from alice keyring: addressMessagingKey 1`] = `
{
  "curve": "ed25519",
  "ecdhDecrypt": [Function],
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      131,
      118,
      45,
      21,
      69,
      60,
      83,
      173,
      4,
      85,
      118,
      136,
      252,
      15,
      202,
      85,
      61,
      79,
      212,
      1,
      42,
      89,
      63,
      240,
      20,
      156,
      229,
      94,
      236,
      12,
      197,
      113,
    ],
    "curve": "ed25519",
  },
  "sign": [Function],
}
`;

exports[`keyring should assert against previous snapshot outputs from alice keyring: rootEncryptionPublicKey 1`] = `
ED25519PublicKey {
  "bytes": Uint8Array [
    48,
    168,
    244,
    4,
    144,
    63,
    142,
    209,
    232,
    152,
    138,
    210,
    72,
    103,
    10,
    204,
    114,
    198,
    47,
    215,
    182,
    116,
    15,
    128,
    200,
    231,
    59,
    252,
    235,
    39,
    192,
    155,
  ],
  "curve": "ed25519",
}
`;

exports[`keyring should assert against previous snapshot outputs from alice keyring: rootInboxKey 1`] = `
ED25519PrivateKey {
  "bytes": Uint8Array [
    184,
    127,
    193,
    149,
    37,
    141,
    126,
    8,
    59,
    5,
    101,
    224,
    182,
    76,
    54,
    71,
    152,
    39,
    21,
    20,
    115,
    165,
    38,
    173,
    1,
    31,
    145,
    226,
    82,
    181,
    113,
    65,
    94,
    48,
    204,
    14,
    58,
    227,
    14,
    149,
    197,
    118,
    96,
    33,
    54,
    70,
    2,
    221,
    74,
    191,
    71,
    233,
    15,
    105,
    24,
    252,
    126,
    55,
    172,
    161,
    100,
    68,
    40,
    14,
  ],
  "curve": "ed25519",
  "keyPair": {
    "publicKey": Uint8Array [
      94,
      48,
      204,
      14,
      58,
      227,
      14,
      149,
      197,
      118,
      96,
      33,
      54,
      70,
      2,
      221,
      74,
      191,
      71,
      233,
      15,
      105,
      24,
      252,
      126,
      55,
      172,
      161,
      100,
      68,
      40,
      14,
    ],
    "secretKey": Uint8Array [
      184,
      127,
      193,
      149,
      37,
      141,
      126,
      8,
      59,
      5,
      101,
      224,
      182,
      76,
      54,
      71,
      152,
      39,
      21,
      20,
      115,
      165,
      38,
      173,
      1,
      31,
      145,
      226,
      82,
      181,
      113,
      65,
      94,
      48,
      204,
      14,
      58,
      227,
      14,
      149,
      197,
      118,
      96,
      33,
      54,
      70,
      2,
      221,
      74,
      191,
      71,
      233,
      15,
      105,
      24,
      252,
      126,
      55,
      172,
      161,
      100,
      68,
      40,
      14,
    ],
  },
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      94,
      48,
      204,
      14,
      58,
      227,
      14,
      149,
      197,
      118,
      96,
      33,
      54,
      70,
      2,
      221,
      74,
      191,
      71,
      233,
      15,
      105,
      24,
      252,
      126,
      55,
      172,
      161,
      100,
      68,
      40,
      14,
    ],
    "curve": "ed25519",
  },
}
`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: accountIdentityKey 1`] = `
{
  "curve": "ed25519",
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      188,
      65,
      178,
      86,
      70,
      133,
      12,
      49,
      224,
      200,
      126,
      190,
      235,
      200,
      89,
      224,
      35,
      135,
      20,
      155,
      81,
      14,
      59,
      96,
      135,
      162,
      233,
      135,
      55,
      208,
      112,
      63,
    ],
    "curve": "ed25519",
  },
  "sign": [Function],
}
`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: accountIdentityKey 2`] = `
Uint8Array [
  66,
  100,
  69,
  26,
  234,
  136,
  32,
  39,
  40,
  243,
  219,
  40,
  72,
  125,
  122,
  185,
  209,
  204,
  14,
  1,
  119,
  33,
  244,
  151,
  16,
  191,
  194,
  244,
  248,
  142,
  243,
  180,
  151,
  74,
  131,
  32,
  232,
  123,
  208,
  248,
  15,
  84,
  15,
  227,
  100,
  221,
  32,
  212,
  103,
  213,
  102,
  57,
  223,
  72,
  129,
  77,
  58,
  26,
  244,
  246,
  225,
  190,
  5,
  9,
]
`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: accountMessagingKey 1`] = `
{
  "curve": "ed25519",
  "ecdhDecrypt": [Function],
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      94,
      99,
      255,
      67,
      91,
      61,
      9,
      82,
      41,
      16,
      34,
      234,
      189,
      191,
      48,
      17,
      9,
      246,
      14,
      69,
      100,
      212,
      77,
      7,
      5,
      193,
      106,
      214,
      104,
      222,
      240,
      35,
    ],
    "curve": "ed25519",
  },
  "sign": [Function],
}
`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: accountMessagingKey 2`] = `Promise {}`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: addressMessagingKey 1`] = `
{
  "curve": "ed25519",
  "ecdhDecrypt": [Function],
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      86,
      185,
      203,
      20,
      37,
      22,
      51,
      26,
      239,
      16,
      25,
      2,
      132,
      7,
      174,
      0,
      234,
      89,
      225,
      165,
      254,
      250,
      111,
      46,
      16,
      233,
      196,
      207,
      92,
      12,
      161,
      239,
    ],
    "curve": "ed25519",
  },
  "sign": [Function],
}
`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: rootEncryptionPublicKey 1`] = `
ED25519PublicKey {
  "bytes": Uint8Array [
    117,
    63,
    208,
    95,
    159,
    228,
    209,
    97,
    121,
    225,
    117,
    51,
    188,
    188,
    67,
    130,
    159,
    154,
    36,
    6,
    253,
    11,
    86,
    232,
    158,
    7,
    136,
    222,
    99,
    244,
    203,
    42,
  ],
  "curve": "ed25519",
}
`;

exports[`keyring should assert against previous snapshot outputs from bob keyring: rootInboxKey 1`] = `
ED25519PrivateKey {
  "bytes": Uint8Array [
    99,
    197,
    118,
    214,
    152,
    75,
    92,
    34,
    32,
    85,
    105,
    72,
    14,
    225,
    152,
    84,
    58,
    77,
    203,
    154,
    205,
    168,
    241,
    14,
    241,
    27,
    106,
    158,
    248,
    193,
    249,
    46,
    248,
    58,
    166,
    64,
    184,
    65,
    136,
    15,
    177,
    212,
    175,
    192,
    19,
    204,
    160,
    201,
    226,
    49,
    207,
    14,
    76,
    74,
    15,
    121,
    25,
    162,
    86,
    226,
    60,
    49,
    178,
    198,
  ],
  "curve": "ed25519",
  "keyPair": {
    "publicKey": Uint8Array [
      248,
      58,
      166,
      64,
      184,
      65,
      136,
      15,
      177,
      212,
      175,
      192,
      19,
      204,
      160,
      201,
      226,
      49,
      207,
      14,
      76,
      74,
      15,
      121,
      25,
      162,
      86,
      226,
      60,
      49,
      178,
      198,
    ],
    "secretKey": Uint8Array [
      99,
      197,
      118,
      214,
      152,
      75,
      92,
      34,
      32,
      85,
      105,
      72,
      14,
      225,
      152,
      84,
      58,
      77,
      203,
      154,
      205,
      168,
      241,
      14,
      241,
      27,
      106,
      158,
      248,
      193,
      249,
      46,
      248,
      58,
      166,
      64,
      184,
      65,
      136,
      15,
      177,
      212,
      175,
      192,
      19,
      204,
      160,
      201,
      226,
      49,
      207,
      14,
      76,
      74,
      15,
      121,
      25,
      162,
      86,
      226,
      60,
      49,
      178,
      198,
    ],
  },
  "publicKey": ED25519PublicKey {
    "bytes": Uint8Array [
      248,
      58,
      166,
      64,
      184,
      65,
      136,
      15,
      177,
      212,
      175,
      192,
      19,
      204,
      160,
      201,
      226,
      49,
      207,
      14,
      76,
      74,
      15,
      121,
      25,
      162,
      86,
      226,
      60,
      49,
      178,
      198,
    ],
    "curve": "ed25519",
  },
}
`;
